src/crypto.ts

```diff
@@ -0,0 +1,361 @@
+/**
+ * crypto
+ *
+ * Architecture
+ * ------------
+ *
+ * The crypto API has two main functions: `getHash` and `getSign`. These functions return functions implementing the
+ * HashFunction and SignFunction interfaces, respectively. HashFunctions and SignFunctions are what user code calls to
+ * hash or sign data.
+ *
+ * Actual implementations are supplied by providers such as `providers/node.ts`. A provider is just a module exporting
+ * the two API functions `getHash` and `getSign`. Since each provider implements the public API, they can be used
+ * directly, although this should not generally be necessary.
+ *
+ * When user code calls this module's `getHash` function, the function immediately starts loading a provider and returns
+ * a wrapper function implementing the HashFunction API. The wrapper will defer all calls until the provider loads. Once
+ * the provider is loaded, the wrapper calls `getHash` on the provider to obtain a real HashFunction instance and
+ * resolves any outstanding calls. Future calls on the wrapper are passed directly to the created HashFunction instance.
+ *
+ * When the HashFunction wrapper's `create` method is called, it returns a wrapped Hasher instance. Like the
+ * HashFunction wrapper, the Hasher wrapper will defer calls until a provider is loaded and a real Hasher implementation
+ * has been created, at which point all future calls to methods on the wrapper are passed directly to the real Hasher.
+ *
+ * The current provider may be requested with the `getProvider` function. Since the initial provider is loaded
+ * asynchronously, this function returns a Promise<CryptoProvider>. The provider may be set using the `setProvider`
+ * function, which accepts a CryptoProvider.
+ */
+
+import Promise from 'dojo-core/Promise';
+import global from 'dojo-core/global';
+import has, { add as hasAdd } from 'dojo-core/has';
+import { Sink } from 'dojo-core/streams/WritableStream';
+
+hasAdd('webcrypto', typeof global.SubtleCrypto !== 'undefined');
```

Comment:
This should be in a dedicated `has` module.  See [dojo-dom/has](https://github.com/dojo/dom/blob/master/src/has.ts) for an example.